File: sources/Android/android-communications/library/src/test/java/com/polar/sdk/api/model/utils/PolarOfflineRecordingUtilsTest.kt
package com.polar.sdk.api.model.utils

import com.google.gson.JsonObject
import com.google.gson.JsonParser
import com.polar.androidcommunications.api.ble.model.gatt.client.psftp.BlePsFtpClient
import com.polar.sdk.impl.utils.PolarOfflineRecordingUtils
import com.polar.sdk.api.model.PolarOfflineRecordingEntry
import io.mockk.mockk
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.asFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.test.runTest
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import java.io.File
import java.io.FileReader
import java.nio.charset.StandardCharsets
import java.time.LocalDateTime

class PolarOfflineRecordingUtilsTest {

    private val mockClient = mockk<BlePsFtpClient>()

    @Test
    fun `listOfflineRecordingsV1 merges split REC files`() = runTest {
        val sampleEntries = listOf(
            Pair("/U/0/20250730/R/101010/ACC0.REC", 500120L),
            Pair("/U/0/20250730/R/101010/ACC1.REC", 500103L),
            Pair("/U/0/20250730/R/101010/ACC2.REC", 102325L),
            Pair("/U/0/20250730/R/101010/HR0.REC", 500000L),
            Pair("/U/0/20250730/R/101010/HR1.REC", 500050L),
            Pair("/U/0/20250730/R/101010/PPG0.REC", 300L)
        )

        val fetchRecursively: (BlePsFtpClient, String, (String) -> Boolean) -> Flow<Pair<String, Long>> =
            { _, _, _ -> sampleEntries.asFlow() }

        val emitted = mutableListOf<PolarOfflineRecordingEntry>()
        val job = launch {
            PolarOfflineRecordingUtils.listOfflineRecordingsV1(mockClient, fetchRecursively)
                .collect { emitted.add(it) }
        }
        job.join()

        val accEntries = emitted.filter { it.path.contains("ACC") }
        val hrEntries = emitted.filter { it.path.contains("HR") }
        val ppgEntries = emitted.filter { it.path.contains("PPG") }

        assert(accEntries.size == 1)
        assert(accEntries[0].size == 500120L + 500103L + 102325L)
        assert(accEntries[0].path.endsWith(".REC"))

        assert(hrEntries.size == 1)
        assert(hrEntries[0].size == 500000L + 500050L)
        assert(hrEntries[0].path.endsWith(".REC"))

        assert(ppgEntries.size == 1)
        assert(ppgEntries[0].size == 300L)
        assert(ppgEntries[0].path.endsWith(".REC"))

    }

    @Test
    fun `listOfflineRecordingsV1 does not return empty files`() = runTest {
        val sampleEntries = listOf(
            Pair("/U/0/20250730/R/101010/ACC0.REC", 500120L),
            Pair("/U/0/20250730/R/101010/ACC1.REC", 500103L),
            Pair("/U/0/20250730/R/101010/ACC2.REC", 0L),
            Pair("/U/0/20250730/R/101010/HR0.REC", 500000L),
            Pair("/U/0/20250730/R/101010/HR1.REC", 0L),
            Pair("/U/0/20250730/R/101010/PPG0.REC", 0L)
        )

        val fetchRecursively: (BlePsFtpClient, String, (String) -> Boolean) -> Flow<Pair<String, Long>> =
            { _, _, _ -> sampleEntries.asFlow() }

        val emitted = mutableListOf<PolarOfflineRecordingEntry>()
        val job = launch {
            PolarOfflineRecordingUtils.listOfflineRecordingsV1(mockClient, fetchRecursively)
                .collect { emitted.add(it) }
        }
        job.join()

        val accEntries = emitted.filter { it.path.contains("ACC") }
        val hrEntries = emitted.filter { it.path.contains("HR") }
        val ppgEntries = emitted.filter { it.path.contains("PPG") }

        assert(accEntries.size == 1)
        assert(accEntries[0].size == 500120L + 500103L)
        assert(accEntries[0].path.endsWith(".REC"))

        assert(hrEntries.size == 1)
        assert(hrEntries[0].size == 500000L)
        assert(hrEntries[0].path.endsWith(".REC"))

        assert(ppgEntries.isEmpty())

    }

    @Test
    fun `listOfflineRecordingsV2 merges split REC files`() = runTest {
        val pmdTxtContent = """
            500120 /U/0/20250730/R/101010/ACC0.REC
            500103 /U/0/20250730/R/101010/ACC1.REC
            102325 /U/0/20250730/R/101010/ACC2.REC
            500000 /U/0/20250730/R/101010/HR0.REC
            500050 /U/0/20250730/R/101010/HR1.REC
            300 /U/0/20250730/R/101010/PPG0.REC
        """.trimIndent().toByteArray(StandardCharsets.UTF_8)

        val emitted = PolarOfflineRecordingUtils.listOfflineRecordingsV2(pmdTxtContent)

        val accEntries = emitted.filter { it.path.contains("ACC") }
        val hrEntries = emitted.filter { it.path.contains("HR") }
        val ppgEntries = emitted.filter { it.path.contains("PPG") }

        assert(accEntries.size == 1)
        assert(accEntries[0].size == 500120L + 500103L + 102325L)
        assert(accEntries[0].path.endsWith(".REC"))

        assert(hrEntries.size == 1)
        assert(hrEntries[0].size == 500000L + 500050L)
        assert(hrEntries[0].path.endsWith(".REC"))

        assert(ppgEntries.size == 1)
        assert(ppgEntries[0].size == 300L)
        assert(ppgEntries[0].path.endsWith(".REC"))

    }

    @Test
    fun `listOfflineRecordingsV2 does not return empty files`() = runTest {
        val pmdTxtContent = """
            500120 /U/0/20250730/R/101010/ACC0.REC
            500103 /U/0/20250730/R/101010/ACC1.REC
            0 /U/0/20250730/R/101010/ACC2.REC
            500050 /U/0/20250730/R/101010/HR0.REC
            0 /U/0/20250730/R/101010/HR1.REC
            0 /U/0/20250730/R/101010/PPG0.REC
        """.trimIndent().toByteArray(StandardCharsets.UTF_8)

        val emitted = PolarOfflineRecordingUtils.listOfflineRecordingsV2(pmdTxtContent)

        val accEntries = emitted.filter { it.path.contains("ACC") }
        val hrEntries = emitted.filter { it.path.contains("HR") }
        val ppgEntries = emitted.filter { it.path.contains("PPG") }

        assert(accEntries.size == 1)
        assert(accEntries[0].size == 500120L + 500103L)
        assert(accEntries[0].path.endsWith(".REC"))

        assert(hrEntries.size == 1)
        assert(hrEntries[0].size == 500050L)
        assert(hrEntries[0].path.endsWith(".REC"))

        assert(ppgEntries.isEmpty())
    }

    @Test
    fun `offline recording PMDFILES golden vectors match Android behavior`() {
        val vector = loadOfflineRecordingVector("pmdfiles-v2-grouping.json")
        val emitted = PolarOfflineRecordingUtils.listOfflineRecordingsV2(
            vector.getAsJsonObject("input").get("pmdFilesTxt").asString.toByteArray(StandardCharsets.UTF_8)
        )

        assertOfflineRecordingEntries(emitted, vector.getAsJsonObject("expected"), platformPathKey = "androidPath")
    }

    @Test
    fun `offline recording golden vectors follow neutral KMP vector shape`() {
        loadOfflineRecordingVectors().forEach { vector ->
            val id = vector.get("id").asString
            assertTrue(id, vector.has("area"))
            assertTrue(id, vector.has("case"))
            assertTrue(id, vector.has("source"))
            assertTrue(id, vector.has("input"))
            assertTrue(id, vector.has("expected"))
            assertTrue(id, vector.has("platforms"))
            val platforms = vector.getAsJsonObject("platforms")
            assertTrue(id, platforms.get("android").asBoolean)
            assertTrue(id, platforms.get("ios").asBoolean)
            assertTrue(id, platforms.get("common").asBoolean)
        }
    }

    @Test
    fun `offline recording metadata readiness manifest is pinned before migration`() {
        val readiness = loadOfflineRecordingVector("metadata-readiness.json")
        val input = readiness.getAsJsonObject("input")
        val expected = readiness.getAsJsonObject("expected")
        val consumerTests = readiness.getAsJsonObject("consumerTests")
        val policyVectorPaths = input.getAsJsonArray("policyVectorPaths").map { it.asString }
        val requiredFamilies = input.getAsJsonArray("requiredBehaviorFamilies").map { it.asString }
        val coveredFamilies = expected.getAsJsonArray("coveredBehaviorFamilies").map { it.asString }

        assertEquals("offline-recording-metadata-readiness", readiness.get("id").asString)
        assertEquals("offlineRecordingMetadataReadiness", input.get("kind").asString)
        assertEquals(
            listOf(
                "sdk/offline-recording/filename-mapping.json",
                "sdk/offline-recording/pmdfiles-v2-grouping.json",
                "sdk/offline-recording/trigger-mapping.json"
            ),
            policyVectorPaths
        )
        val expectedFamilies = listOf(
            "filename-to-measurement-type-mapping",
            "split-file-index-stripping",
            "invalid-filename-boundary",
            "pmdfiles-grouping",
            "zero-size-recording-filtering",
            "invalid-entry-filtering",
            "representative-path-platform-policy",
            "trigger-model-projection",
            "disabled-trigger-filtering",
            "platform-offline-recording-vector-reference-gate",
            "compile-verification-gate"
        )
        assertEquals(expectedFamilies, requiredFamilies)
        assertEquals(expectedFamilies, coveredFamilies)
        assertEquals(
            "Offline recording metadata migration may proceed only after every vector named by this readiness manifest is executable from shared commonTest, Android and iOS metadata tests continue to reference the same vectors, filename classification, split-file normalization, invalid filename handling, PMDFILES grouping, zero-size and invalid-entry filtering, representative path policy, trigger model projection, disabled-trigger filtering, and compile verification remain explicit before production metadata mapping moves.",
            expected.get("commonDecision").asString
        )
        assertEquals(
            listOf(
                "com.polar.androidcommunications.api.ble.model.offlinerecording.OfflineRecordingUtilityTest",
                "com.polar.sdk.impl.utils.PolarDataUtilsTest",
                "com.polar.sdk.api.model.utils.PolarOfflineRecordingUtilsTest"
            ),
            consumerTests.getAsJsonArray("android").map { it.asString }
        )
        assertEquals(
            listOf(
                "OfflineRecordingUtilsTest",
                "PolarDataUtilsTest",
                "PolarOfflineRecordingUtilsTest"
            ),
            consumerTests.getAsJsonArray("ios").map { it.asString }
        )
        assertEquals(
            listOf("com.polar.sharedtest.OfflineRecordingMetadataCommonPolicyTest"),
            consumerTests.getAsJsonArray("commonPrototype").map { it.asString }
        )
    }

    private fun assertOfflineRecordingEntries(actual: List<PolarOfflineRecordingEntry>, expected: JsonObject, platformPathKey: String) {
        val expectedEntries = expected.getAsJsonArray("entries").map { it.asJsonObject }
        assertEquals("entry count", expectedEntries.size, actual.size)
        expectedEntries.forEach { expectedEntry ->
            val actualEntry = actual.firstOrNull { it.type.name == expectedEntry.get("type").asString } ?: error("Missing ${expectedEntry.get("type").asString}")
            assertEquals(expectedEntry.get(platformPathKey).asString, actualEntry.path)
            assertEquals(expectedEntry.get("size").asLong, actualEntry.size)
            assertEquals(LocalDateTime.parse(expectedEntry.get("dateTime").asString), actualEntry.date)
        }
    }

    private fun loadOfflineRecordingVector(fileName: String): JsonObject {
        FileReader(findRepositoryRoot().resolve("testdata/golden-vectors/sdk/offline-recording/$fileName")).use { reader ->
            return JsonParser().parse(reader).asJsonObject
        }
    }

    private fun loadOfflineRecordingVectors(): List<JsonObject> {
        val vectorDirectory = findRepositoryRoot()
            .resolve("testdata/golden-vectors/sdk/offline-recording")
        return vectorDirectory
            .listFiles { file -> file.isFile && file.extension == "json" }
            .orEmpty()
            .sortedBy { it.name }
            .map { file ->
                FileReader(file).use { reader ->
                    JsonParser().parse(reader).asJsonObject
                }
            }
            .filterNot { vector -> vector.getAsJsonObject("input")?.get("kind")?.asString == "offlineRecordingMetadataReadiness" }
    }

    private fun findRepositoryRoot(): File {
        val userDirectory = System.getProperty("user.dir") ?: error("user.dir is not set")
        var directory = File(userDirectory).absoluteFile
        while (true) {
            if (directory.resolve("testdata/golden-vectors").isDirectory) {
                return directory
            }
            directory = directory.parentFile ?: error("Could not find repository root from $userDirectory")
        }
    }
}

File: sources/iOS/ios-communications/Tests/PolarBleSdkTests/PolarOfflineRecordingUtilsTest.swift
//  Copyright © 2025 Polar. All rights reserved.

import XCTest
@testable import PolarBleSdk

final class PolarOfflineRecordingUtilsTest: XCTestCase {

    private var mockClient: MockBlePsFtpClient!
    
    override func setUpWithError() throws {
        mockClient = MockBlePsFtpClient(gattServiceTransmitter: MockPolarGattServiceTransmitter())
    }

    override func tearDownWithError() throws {
        mockClient = nil
    }

    func testListOfflineRecordingsV1_mergesSplitRecFiles() async throws {
        // Arrange
        let sampleEntries: [(String, UInt)] = [
            ("/U/0/20250730/R/101010/ACC0.REC", 500_120),
            ("/U/0/20250730/R/101010/ACC1.REC", 500_103),
            ("/U/0/20250730/R/101010/ACC2.REC", 102_325),
            ("/U/0/20250730/R/101010/HR0.REC", 500_000),
            ("/U/0/20250730/R/101010/HR1.REC", 500_050),
            ("/U/0/20250730/R/101010/PPG0.REC", 300)
        ]
        let fetchRecursively: (BlePsFtpClient, String, @escaping (String) -> Bool) async throws -> [(String, UInt)] = { _, _, _ in
            sampleEntries
        }

        // Act
        let emitted = try await collectStream(PolarOfflineRecordingUtils.listOfflineRecordingsV1(
            client: mockClient,
            fetchRecursively: fetchRecursively
        ))

        // Assert
        let accEntries = emitted.filter { $0.path.contains("ACC") }
        let hrEntries  = emitted.filter { $0.path.contains("HR") }
        let ppgEntries = emitted.filter { $0.path.contains("PPG") }

        XCTAssertEqual(accEntries.count, 1)
        XCTAssertEqual(accEntries.first?.size, 500_120 + 500_103 + 102_325)
        XCTAssertTrue(accEntries.first?.path.hasSuffix(".REC") ?? false)

        XCTAssertEqual(hrEntries.count, 1)
        XCTAssertEqual(hrEntries.first?.size, 500_000 + 500_050)
        XCTAssertTrue(hrEntries.first?.path.hasSuffix(".REC") ?? false)

        XCTAssertEqual(ppgEntries.count, 1)
        XCTAssertEqual(ppgEntries.first?.size, 300)
        XCTAssertTrue(ppgEntries.first?.path.hasSuffix(".REC") ?? false)

        emitted.forEach { XCTAssertNotNil($0.date) }
    }

    func testListOfflineRecordingsV1_returns_empty_when_regex_fails() async throws {
        // Arrange
        let sampleEntries: [(String, UInt)] = [("/U/0/2025073/R/101010/PPG0.REC", 300)]
        let fetchRecursively: (BlePsFtpClient, String, @escaping (String) -> Bool) async throws -> [(String, UInt)] = { _, _, _ in
            sampleEntries
        }

        // Act
        let emitted = try await collectStream(PolarOfflineRecordingUtils.listOfflineRecordingsV1(
            client: mockClient, fetchRecursively: fetchRecursively))

        // Assert
        XCTAssertEqual(emitted.filter { $0.path.contains("PPG") }.count, 0)
    }

    func testListOfflineRecordingsV1_returns_empty_when_date_parsing_fails() async throws {
        // Arrange
        let sampleEntries: [(String, UInt)] = [("/U/0/99999999/R/101010/PPG0.REC", 300)]
        let fetchRecursively: (BlePsFtpClient, String, @escaping (String) -> Bool) async throws -> [(String, UInt)] = { _, _, _ in
            sampleEntries
        }

        // Act
        let emitted = try await collectStream(PolarOfflineRecordingUtils.listOfflineRecordingsV1(
            client: mockClient, fetchRecursively: fetchRecursively))

        // Assert
        XCTAssertEqual(emitted.filter { $0.path.contains("PPG") }.count, 0)
    }

    func testListOfflineRecordingsV1_returns_empty_when_time_parsing_fails() async throws {
        // Arrange
        let sampleEntries: [(String, UInt)] = [("/U/0/20250730/R/999999/PPG0.REC", 300)]
        let fetchRecursively: (BlePsFtpClient, String, @escaping (String) -> Bool) async throws -> [(String, UInt)] = { _, _, _ in
            sampleEntries
        }

        // Act
        let emitted = try await collectStream(PolarOfflineRecordingUtils.listOfflineRecordingsV1(
            client: mockClient, fetchRecursively: fetchRecursively))

        // Assert
        XCTAssertEqual(emitted.filter { $0.path.contains("PPG") }.count, 0)
    }

    func testListOfflineRecordingsV1_returns_empty_when_meas_type_parsing_fails() async throws {
        // Arrange
        let sampleEntries: [(String, UInt)] = [("/U/0/20250730/R/101010/ZZZ9.REC", 300)]
        let fetchRecursively: (BlePsFtpClient, String, @escaping (String) -> Bool) async throws -> [(String, UInt)] = { _, _, _ in
            sampleEntries
        }

        // Act
        let emitted = try await collectStream(PolarOfflineRecordingUtils.listOfflineRecordingsV1(
            client: mockClient, fetchRecursively: fetchRecursively))

        // Assert
        XCTAssertEqual(emitted.filter { $0.path.contains("PPG") }.count, 0)
    }

    func testListOfflineRecordingsV1_returns_empty_when_empty_file() async throws {
        // Arrange
        let sampleEntries: [(String, UInt)] = [
            ("/U/0/20250730/R/101010/ACC0.REC", 500_120),
            ("/U/0/20250730/R/101010/ACC1.REC", 500_103),
            ("/U/0/20250730/R/101010/ACC2.REC", 0),
            ("/U/0/20250730/R/101010/HR0.REC", 500_000),
            ("/U/0/20250730/R/101010/HR1.REC", 0),
            ("/U/0/20250730/R/101010/PPG0.REC", 0)
        ]
        let fetchRecursively: (BlePsFtpClient, String, @escaping (String) -> Bool) async throws -> [(String, UInt)] = { _, _, _ in
            sampleEntries
        }

        // Act
        let emitted = try await collectStream(PolarOfflineRecordingUtils.listOfflineRecordingsV1(
            client: mockClient, fetchRecursively: fetchRecursively))

        // Assert
        let accEntries = emitted.filter { $0.path.contains("ACC") }
        let hrEntries  = emitted.filter { $0.path.contains("HR") }
        let ppgEntries = emitted.filter { $0.path.contains("PPG") }

        XCTAssertEqual(accEntries.count, 1)
        XCTAssertEqual(accEntries.first?.size, 500_120 + 500_103)
        XCTAssertTrue(accEntries.first?.path.hasSuffix(".REC") ?? false)

        XCTAssertEqual(hrEntries.count, 1)
        XCTAssertEqual(hrEntries.first?.size, 500_000)
        XCTAssertTrue(hrEntries.first?.path.hasSuffix(".REC") ?? false)

        XCTAssertEqual(ppgEntries.count, 0)

        emitted.forEach { XCTAssertNotNil($0.date) }
    }

    func testListOfflineRecordingsV2_mergesSplitRecFiles() throws {
        // Arrange
        let pmdTxt = """
        500120 /U/0/20250730/R/101010/ACC0.REC
        500103 /U/0/20250730/R/101010/ACC1.REC
        102325 /U/0/20250730/R/101010/ACC2.REC
        500000 /U/0/20250730/R/101010/HR0.REC
        500050 /U/0/20250730/R/101010/HR1.REC
        300 /U/0/20250730/R/101010/PPG0.REC
        """.data(using: .utf8)!

        // Act
        let emitted = try PolarOfflineRecordingUtils.listOfflineRecordingsV2(fileData: pmdTxt)

        // Assert
        let accEntries = emitted.filter { $0.path.contains("ACC") }
        let hrEntries  = emitted.filter { $0.path.contains("HR") }
        let ppgEntries = emitted.filter { $0.path.contains("PPG") }

        XCTAssertEqual(accEntries.count, 1)
        XCTAssertEqual(accEntries.first?.size, 500_120 + 500_103 + 102_325)
        XCTAssertTrue(accEntries.first?.path.hasSuffix(".REC") ?? false)

        XCTAssertEqual(hrEntries.count, 1)
        XCTAssertEqual(hrEntries.first?.size, 500_000 + 500_050)
        XCTAssertTrue(hrEntries.first?.path.hasSuffix(".REC") ?? false)

        XCTAssertEqual(ppgEntries.count, 1)
        XCTAssertEqual(ppgEntries.first?.size, 300)
        XCTAssertTrue(ppgEntries.first?.path.hasSuffix(".REC") ?? false)

        emitted.forEach { XCTAssertNotNil($0.date) }
    }

    func testListOfflineRecordingsV2_returns_empty_when_empty_file() throws {
        // Arrange
        let pmdTxt = """
        500120 /U/0/20250730/R/101010/ACC0.REC
        500103 /U/0/20250730/R/101010/ACC1.REC
        0 /U/0/20250730/R/101010/ACC2.REC
        500000 /U/0/20250730/R/101010/HR0.REC
        0 /U/0/20250730/R/101010/HR1.REC
        0 /U/0/20250730/R/101010/PPG0.REC
        """.data(using: .utf8)!

        // Act
        let emitted = try PolarOfflineRecordingUtils.listOfflineRecordingsV2(fileData: pmdTxt)

        // Assert
        let accEntries = emitted.filter { $0.path.contains("ACC") }
        let hrEntries  = emitted.filter { $0.path.contains("HR") }
        let ppgEntries = emitted.filter { $0.path.contains("PPG") }

        XCTAssertEqual(accEntries.count, 1)
        XCTAssertEqual(accEntries.first?.size, 500_120 + 500_103)
        XCTAssertTrue(accEntries.first?.path.hasSuffix(".REC") ?? false)

        XCTAssertEqual(hrEntries.count, 1)
        XCTAssertEqual(hrEntries.first?.size, 500_000)
        XCTAssertTrue(hrEntries.first?.path.hasSuffix(".REC") ?? false)

        XCTAssertEqual(ppgEntries.count, 0)

        emitted.forEach { XCTAssertNotNil($0.date) }
    }

    func testOfflineRecordingPmdFilesGoldenVectorsMatchIOSBehavior() throws {
        let vector = try loadOfflineRecordingVector("pmdfiles-v2-grouping.json")
        let input = try XCTUnwrap(vector["input"] as? [String: Any])
        let pmdFilesTxt = try XCTUnwrap(input["pmdFilesTxt"] as? String)
        let emitted = try PolarOfflineRecordingUtils.listOfflineRecordingsV2(fileData: try XCTUnwrap(pmdFilesTxt.data(using: .utf8)))

        try assertOfflineRecordingEntries(emitted, expected: try XCTUnwrap(vector["expected"] as? [String: Any]), platformPathKey: "iosPath")
    }

    func testOfflineRecordingPmdFilesGoldenVectorsFollowNeutralKmpVectorShape() throws {
        try assertNeutralKmpVectorShape(try loadOfflineRecordingVector("pmdfiles-v2-grouping.json"), id: "pmdfiles-v2-grouping.json")
    }

    func testOfflineRecordingMetadataReadinessManifestPinsSharedOwnership() throws {
        let readiness = try loadOfflineRecordingVector("metadata-readiness.json")
        let input = try XCTUnwrap(readiness["input"] as? [String: Any])
        let expected = try XCTUnwrap(readiness["expected"] as? [String: Any])
        let consumerTests = try XCTUnwrap(readiness["consumerTests"] as? [String: Any])
        let platforms = try XCTUnwrap(readiness["platforms"] as? [String: Any])
        let policyVectorPaths = try XCTUnwrap(input["policyVectorPaths"] as? [String])
        let requiredFamilies = try XCTUnwrap(input["requiredBehaviorFamilies"] as? [String])
        let coveredFamilies = try XCTUnwrap(expected["coveredBehaviorFamilies"] as? [String])
        XCTAssertEqual(readiness["id"] as? String, "offline-recording-metadata-readiness")
        XCTAssertEqual(input["kind"] as? String, "offlineRecordingMetadataReadiness")
        XCTAssertEqual(policyVectorPaths, [
            "sdk/offline-recording/filename-mapping.json",
            "sdk/offline-recording/pmdfiles-v2-grouping.json",
            "sdk/offline-recording/trigger-mapping.json"
        ])
        let expectedFamilies = [
            "filename-to-measurement-type-mapping",
            "split-file-index-stripping",
            "invalid-filename-boundary",
            "pmdfiles-grouping",
            "zero-size-recording-filtering",
            "invalid-entry-filtering",
            "representative-path-platform-policy",
            "trigger-model-projection",
            "disabled-trigger-filtering",
            "platform-offline-recording-vector-reference-gate",
            "compile-verification-gate"
        ]
        XCTAssertEqual(requiredFamilies, expectedFamilies)
        XCTAssertEqual(coveredFamilies, expectedFamilies)
        XCTAssertEqual(expected["commonDecision"] as? String, "Offline recording metadata shared ownership remains valid while every vector named by this readiness manifest is executable from shared commonTest, Android and iOS metadata tests continue to reference the same vectors, filename classification, split-file normalization, invalid filename handling, PMDFILES grouping, zero-size and invalid-entry filtering, representative path policy, trigger model projection, disabled-trigger filtering, and compile verification remain explicit before production metadata mapping moves.")
        XCTAssertEqual(try XCTUnwrap(consumerTests["android"] as? [String]), [
            "com.polar.androidcommunications.api.ble.model.offlinerecording.OfflineRecordingUtilityTest",
            "com.polar.sdk.impl.utils.PolarDataUtilsTest",
            "com.polar.sdk.api.model.utils.PolarOfflineRecordingUtilsTest"
        ])
        XCTAssertEqual(try XCTUnwrap(consumerTests["ios"] as? [String]), [
            "OfflineRecordingUtilsTest",
            "PolarDataUtilsTest",
            "PolarOfflineRecordingUtilsTest"
        ])
        XCTAssertEqual(try XCTUnwrap(consumerTests["commonPrototype"] as? [String]), ["com.polar.sharedtest.OfflineRecordingMetadataCommonPolicyTest"])
        XCTAssertEqual(platforms["android"] as? Bool, true)
        XCTAssertEqual(platforms["ios"] as? Bool, true)
        XCTAssertEqual(platforms["common"] as? Bool, true)
    }

    // MARK: - Helpers

    private func collectStream<T>(_ stream: AsyncThrowingStream<T, Error>) async throws -> [T] {
        var results: [T] = []
        for try await value in stream { results.append(value) }
        return results
    }

    private func assertOfflineRecordingEntries(_ actual: [PolarOfflineRecordingEntry], expected: [String: Any], platformPathKey: String) throws {
        let expectedEntries = try XCTUnwrap(expected["entries"] as? [[String: Any]])
        XCTAssertEqual(actual.count, expectedEntries.count, "entry count")
        for expectedEntry in expectedEntries {
            let expectedType = try polarDeviceDataType(named: try XCTUnwrap(expectedEntry["type"] as? String))
            let actualEntry = try XCTUnwrap(actual.first { $0.type == expectedType }, "\(expectedType)")
            XCTAssertEqual(actualEntry.path, try XCTUnwrap(expectedEntry[platformPathKey] as? String))
            XCTAssertEqual(actualEntry.size, UInt(try number(expectedEntry, "size")))
            try assertDate(actualEntry.date, expected: try XCTUnwrap(expectedEntry["dateTime"] as? String))
        }
    }

    private func polarDeviceDataType(named name: String) throws -> PolarDeviceDataType {
        switch name {
        case "ACC": return .acc
        case "GYRO": return .gyro
        case "MAGNETOMETER": return .magnetometer
        case "PPG": return .ppg
        case "PPI": return .ppi
        case "HR": return .hr
        case "TEMPERATURE": return .temperature
        case "SKIN_TEMPERATURE": return .skinTemperature
        default: throw NSError(domain: "PolarOfflineRecordingUtilsTest", code: 2, userInfo: [NSLocalizedDescriptionKey: "Unknown device data type \(name)"])
        }
    }

    private func assertDate(_ actual: Date, expected: String) throws {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss"
        formatter.locale = Locale(identifier: "en_US_POSIX")
        XCTAssertEqual(formatter.string(from: actual), expected)
    }

    private func number(_ object: [String: Any], _ key: String) throws -> Int {
        return try XCTUnwrap(object[key] as? NSNumber, key).intValue
    }

    private func assertNeutralKmpVectorShape(_ vector: [String: Any], id: String) throws {
        XCTAssertNotNil(vector["area"], id)
        XCTAssertNotNil(vector["case"], id)
        XCTAssertNotNil(vector["source"], id)
        XCTAssertNotNil(vector["input"], id)
        XCTAssertNotNil(vector["expected"], id)
        let platforms = try XCTUnwrap(vector["platforms"] as? [String: Any], id)
        XCTAssertNotNil(platforms["android"], id)
        XCTAssertNotNil(platforms["ios"], id)
        XCTAssertNotNil(platforms["common"], id)
    }

    private func loadOfflineRecordingVector(_ fileName: String) throws -> [String: Any] {
        let file = try GoldenVectorTestData.repositoryRoot()
            .appendingPathComponent("testdata/golden-vectors/sdk/offline-recording")
            .appendingPathComponent(fileName)
        let data = try Data(contentsOf: file)
        return try XCTUnwrap(JSONSerialization.jsonObject(with: data) as? [String: Any], file.path)
    }

}
